_posts/2020-05-19-docker-rust-notpresent.md

```diff
@@ -0,0 +1,48 @@
+---
+title: "Debugging Error: NotPresent with Rust in Docker"
+description: "A friend's Docker container was exiting with Error: NotPresent and no other information. I embarked on a deep dive for an ultimately simple cause."
+layout: post
+date: 2020-05-19
+image: https://assets.csh.rit.edu/pubsite/projects/carbon.png
+image-sm: https://assets.csh.rit.edu/pubsite/projects/carbon.png
+author: Galen Guyer
+author-image: https://cdn.galenguyer.com/img/avatars/headshot.jpg
+author-email: chef@csh.rit.edu
+author-social:
+    github: https://github.com/galenguyer/
+categories: [docker, rust]
+---
+
+<a href='https://galenguyer.com/blog/2020/05/19/docker-rust-notpresent'>Check out this post on Galen's blog!</a>
+
+<p>Today, a friend posted the following snippet in Slack:</p>
+<pre> [~/Documents/shelflife] [±docker✗] $ docker run -v -it --rm --name my-running-app shelflife-rust-docker
+    Error: NotPresent</pre>
+<p>The first guess was they deleted the docker image and forgot to rebuild it. That wasn't the case, leading me to dive into this cryptic error message that we assumed some some hidden Rust or Docker error. We weren't able to find any example of this error message for Rust or Docker online, so I knew this would be a fun challenge.</p>
+
+<p>The first thing I did was to make sure the path to the binary was working, it'd be silly if this wasn't working because <code>/usr/local/bin</code> wasn't in the container's PATH. I changed the last line in the Dockerfile from <code>CMD ["shelflife"]</code> to <code>ENTRYPOINT ["/usr/local/bin/shelflife"]</code>. The error persisted, which means it wasn't a PATH error. Of course this couldn't be easy.</p>
```

Comment:
```suggestion
<p>The first thing I did was to make sure the path to the binary was working, it'd be silly if this wasn't working because <code>/usr/local/bin</code> wasn't in the container's PATH. I changed the last line in the Dockerfile from <code>CMD ["shelflife"]</code> to <code>ENTRYPOINT ["/usr/local/bin/shelflife"]</code>. The error persisted, which meant it wasn't a PATH error. Of course this couldn't be easy.</p>
```